docs: Add scripts overview to readme (#286)
superfaceai/station: README.md

# Run tests
$ yarn test

# Record new trafic with live API calls, more information bellow
$ yarn test:record grid/path/to/test.ts
```

> Jest in station is configured to run tests inside `/grid` folder by default, with reporter located in `/jest/reporter` and can use groups to filter which tests to run (mainly used for tagging tests ready to continuous testing).

Author: I wanted to add overview of test configurations that station has/will have, since with one I added in #263 there would be 3 `jest.config.js` files, but totally failed describing it 😄 

I can remove this and describe it under relevant commands and later add this section after #263 is merged 🤷🏻‍♂️

> Jest in station is configured to run tests inside `/grid` folder by default, with reporter located in `/jest/reporter` and can use groups to filter which tests to run (mainly used for tagging tests ready to continuous testing).

Reviewer: Yeah, Ondra explained to me that this section will be moved to the contribution guide, so it probably makes more sense there. I'd dedicate a separate section there to configuration files.

> Jest in station is configured to run tests inside `/grid` folder by default, with reporter located in `/jest/reporter` and can use groups to filter which tests to run (mainly used for tagging tests ready to continuous testing).

Author: Ok, I'll remove this with example

```javascript
module.exports = {
  preset: 'ts-jest',
  testEnvironment: 'node',
  rootDir: './grid',
  runner: 'groups',
  reporters: ['default', '../dist/reporter.js']
};
```

> ### **Test with mocked traffic**

Reviewer: nitpick: Could you please get rid of these bolds around headings?

```shell
$ yarn test
```

This command runs tests with mocked traffic and therefore no live traffic is enabled. You can find recordings that this command uses as mocked traffic in `/nock`.

Recordings are categorised similarly as integrations in `/grid` folder:
`nock/<scope>/<profile-name>/<provider-name>/<usecase-name>/recording-<hash>.json`

- every recording gets hashed based on **input**, **test name** or **custom test name**
- this hash can be only seen with use of `DEBUG=superface:testing:hash*`

> 

Author: They're used individually, I guess it would make sense to move it elsewhere since it's relevant to all commands.

### Test with live traffic

```shell
$ yarn test:record
```

With this command, live traffic is enabled and therefore maps send requests to the live providers. This command also records this request and response. There are multiple scenarios what can happen:

> - use-case does not have recording → records traffic and saves as default one
> - use-case already have recording → records new traffic and compares it to the old one →
>     - if it matches → do not save recording
>     - if it doesn’t match → save new recording next to old one with suffix `-new`

Reviewer: ```suggestion
- If a use case doesn't have recording → save it as a default recording
- If a use case has recording → record new traffic and compare it to the old one →
    - If the recordings match → don't save the new recording
    - If the recordings don’t match → save the new recording next to the old one with suffix `-new`
```

### Test with new mocked traffic

```shell
$ yarn test:with_new_traffic
```

> This command run tests without live traffic, similar as `yarn test`, but uses new recordings instead of default ones (if present). New recordings are always newly recorded traffic that does not match the default old one and are located next to the old one with suffix `-new`.

Reviewer: question: What happens if a new recording is not present? Is the default one used or is the test ignored or...?

```suggestion
Similar to `yarn test`, run tests against a mock server with prerecorded traffic, but use any available new recordings instead of default ones. New recordings are created by `yarn test:record` when the newly recorded traffic doesn't match the old one. New recordings are located next to old ones with suffix `new`.
```

### Test and replace old recording with new one

> **⚠️ This command should be run only when you’re sure that tests are passing with new recordings → `yarn test:with_new_traffic`**

Reviewer: tip: Here's a [neat trick for callouts](https://github.com/community/community/discussions/16925):

```suggestion
> **Warning**
> Run this command only when you're sure the tests pass with the new recordings (using `yarn test:with_new_traffic`).
```

```shell
$ yarn test:update_traffic
```

> This commands run tests with mocked traffic, but also replaces old recording with new one (if present) before loading recordings - this means that tests are run with mocked new traffic already.

Reviewer: ```suggestion
Replace the old traffic recordings with new ones (if present) and run tests using these new recordings.
```

### Test with live traffic with development reporting

```shell
$ yarn test:record:dev
```

> This command run tests with live traffic and record it, similar as `yarn test:record`. The script implementation looks like this:
> 
> ```shell
> $ TEST_ENV='dev' yarn test:record --group=live/safe
> ```
> 
> `TEST_ENV` sets development environment for reporter - to use local reporter that just logs the provider changes into console
> 
> `--group=live/safe` filters what tests to run, thanks to [jest-runner-groups](https://www.npmjs.com/package/jest-runner-groups) 

Reviewer: ```suggestion
Run tests under `live/safe` tests group with live traffic (similar to `yarn test:record`) and console reporter. This is useful for development of continuous live tests.
```

### Test with live traffic with production reporting

```shell
$ yarn test:record:prod
```

> This command run tests with live traffic and record it, similar as `yarn test:record` or `yarn test:record:dev`. Only difference between `test:record:dev` is that it uses production reporter that reports provider changes to the public slack channel.
> 
> Slack channel can be changed with env variable `PROD_REPORTING_DESTINATION`. It expects channel id.

Reviewer: Ooof, suddenly you drop Slack reporting on me. Like you tell me how to change the Slack channel ID without any prior info on how to setup the reporter, Slack bot, access tokens etc. I don't think this should be documented here, but in a separate document, but even then it will be probably only for us.

```suggestion
Run tests under `live/safe` group with live traffic, record it and report provider changes to Slack.

Set Slack channel ID with env variable `PROD_REPORTING_DESTINATION`.
```

Author: Good point, some section about setting up reporter would be good.

> ### Test and update snapshots

Reviewer: ```suggestion
### Updating test snapshots
```

> Since lot of tests written in station uses jest snapshots for comparing values from map, you might need to update snapshots to modify/add/remove some of them. You can do this simply by adding option `-u` or `--updateSnapshot`. For example:
> 
> ```shell
> $ yarn test -u
> 
> $ yarn test:record --updateSnapshot
> ```

Reviewer: - Do I need to always use `--updateSnapshot`? Previously it was included with `test:record`
- _simply_ (sighs)
- Don't explain snapshots to me, but if you want to, link me to Jest docs (too lazy to look for the link now)

````suggestion
If you change the map or record a new response, you may need to update Jest snapshots captured during the previous run. Use `-u` parameter to update snapshots:

```shell
$ yarn test -u

$ yarn test:record -u
```
````

Author: Yes, it was included in there, but we decided to force passing tests when recording and therefore it would overwrite them (see [here](https://github.com/superfaceai/station/blob/9a476018e82b7f8211f6454f46bb4dd2f0b45499/jest/setup.config.js))

---

> ### Overall filtering of tests

Reviewer: ```suggestion
### Tests filtering
```

> In each of command described above, you can use argument to specify integration, for example:

Reviewer: This doesn't explain what I should enter here. Is it a directory path or use case identifier (with scope and profile an provider tacked on)?

Also, can I use a glob (like `chat/*/maps/slack`)?

```shell
$ yarn test chat/messages/maps/slack

$ yarn test:record chat/messages/maps/slack

$ TEST_ENV='dev' yarn test:record chat/messages
```

> This is important mainly in case of testing with live traffic when new recordings are stored and compared with old ones for each test that you run. For example if you want to record new traffic just for slack provider in profile `chat/messages`, you don’t want to record new traffic also for other providers in this profile.

Reviewer: Move this paragraph above the code block above, to continue with the explained example.

```suggestion
This is necessary whenever you record a traffic to limit interaction with live APIs only to the changed maps and tests.

For example, if you want to record new traffic just for the `slack` provider and profile `chat/messages`, and ignore other providers:
```

### Testing library DEBUG

> It also helps when you use `DEBUG` environment variable to see logs during test run.

Reviewer: ```suggestion
Use `DEBUG` environment variable to check the testing library behavior during a test run.
```

- log **everything**: `DEBUG=superface:testing*`
- log **main setup**: `DEBUG=superface:testing`
    - perform results
    - start and end of recording/mocking HTTP traffic
    - start of `beforeRecordingLoad` and `beforeRecordingSave` functions
- log **setup** of superface components and recordings: `DEBUG=superface:testing:setup*`
    - setup of recording paths and superface components (profile, provider, usecase)
    - setup of superjson and local map
- log **hashing** of recordings: `DEBUG=superface:testing:hash*`
    - creation of recording hash
- log **recording/replacing** information: `DEBUG=superface:testing:recordings*`
- log **matching** information: `DEBUG=superface:testing:matching*`
- log **reporting** information:  `DEBUG=superface:testing:reporter*`

> ### Network errors

Reviewer: ```suggestion
### Troubleshooting common errors
```

> `NetworkError: Fetch failed: reject issue`
> 
> `SdkExecutionError: Request ended with network error: reject`
> 
> If you get `NetworkError` or `SdkExecutionError` during testing with mocked traffic(`yarn test`), it usually means that request didn’t get through. If nock (used for loading mocked traffic) can’t match recording, request is denied.
> 
> You can debug nock matching of recordings with `DEBUG=nock*` to see what went wrong.

Reviewer: ````suggestion
You can run into the following errors when running tests:

```
NetworkError: Fetch failed: reject issue
SdkExecutionError: Request ended with network error: reject
```

When you are testing with mocked traffic (using `yarn test`), it means that the request didn't match any recordings and was rejected. Try rerecording the traffic using `yarn test:record`. You can also debug the matching behavior with `DEBUG=nock*` env variable, 
````
